GitUI/CommandsDialogs/FormCommit.cs

```diff
@@ -192,6 +195,7 @@ public FormCommit(GitUICommands aCommands, CommitKind commitKind, GitRevision ed
             Unstaged.FilterVisible = true;
             Staged.SetNoFilesText(_noStagedChanges.Text);
 
+            _formattedCommitMessage = new FormattedCommitMessage(_commentChar);
```

Comment:
As I stated before I have reservations about the proposed API. I find them unintuitive. They are also not following SOLID principles, which contributes to techdebt and degrades maintenability of the code.
There has to be a clear separation of state from behavior, otherwise it makes it difficult to reason about the code, substitute behaviors or process different data.


`FormattedCommitMessage` reads as though it represents a DTO/POCO, however it then used as a behavioral component. This is confusing.

You may have started the implementation from the actual formatter, let's try to reverse the situation and consider how the formatter should be used first.
As a consumer and your fellow developer I would expect the consumption API to read something like the following:

```
private readonly ICommitMessageFormatter _commitMessageFormatter;

public FormCommit(...)
{
    // ...

    var commentChar = Module.EffectiveConfigFile.GetString(SettingKeyString.CommentChar, "#");
    _commitMessageFormatter = new CommitMessageFormatter(commentChar);

    // ...
}


private void FormatAllText()
{
    var formattedMessage = _commitMessageFormatter.Format(Message.GetLines(), Message.CurrentLine - 1, Message.CurrentColumn - 1);
    // ...
}

```

The above reads almost like a natural language. It is also reasonably clear what how the input translates into the output. The cursor location isn't very clear, but that's a separate issue.

I would expect the `formattedMessage` to be a DTO/POCO (i.e. no behaviors, just state). I would also expect it to contains all necessary information for me render the result to the user.

```
if (formattedMessage.RequiresFullUpdate)
{
    var formattedLines = formattedMessage.FormattedLines.ToArray();
    Message.SetLines(formattedLines, formattedMessage.CursorLine, formattedMessage.CursorColumn);
}
foreach (var coloredLine in formattedMessage.ColoredLines)
{
    Message.ChangeTextColor(coloredLine.LineNumber, coloredLine.From, coloredLine.Length, Color.Red);
}
```

Based on how I expect the formatter API to be used, the formatter will have a pretty simple interface:

```
public interface ICommitMessageFormatter 
{
    FormattedMessage Format(IEnumerable<string> lines, int cursorLine, int cursorColumn);
}
```


I hope it makes sense.

Reply:
I'm not sure I quite agree here...

> There has to be a clear separation of state from behavior, otherwise it makes it difficult to reason about the code, substitute behaviors or process different data.

What about encapsulation ? On the contrary, having state an behavior separated introduces coupling between the representations of both.

> `FormattedCommitMessage` reads as though it represents a DTO/POCO, however it then used as a behavioral component. This is confusing.

I don't see what makes you feel this way : sure a DTO could be named this way, but here the main objective is to avoid having a behemoth class such as `FormCommit` by extracting part of its existing implementation into a self-contained, encapsulated (and testable) class. As a standard OO class, it maintains its state and how to manipulate it together.

On the other hand, an `ICommitMessageFormatter` shouldn't maintain state, so how could it keep internal information for instance for optimization purposes ? The DTO could keep such internal data specific to an implementation, possibly in a generic way, but it seems to get quickly complicated. Moreover, I'm really not sure there is a significant chance for a completely different formatting behavior that would warrant an alternative implementation rather than adding a few configuration options.

Would it be clearer to get rid of the "Formatting" words ? The class would be just `CommitMessage`, with simple `Lines`, and `FormCommit.FormatAllText()` renamed `UpdateCommitMessage()`.